### PR TITLE
types and conversions for porter app def (#3378)

### PR DESCRIPTION
**Pull request by ianedwards**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

### CODE REVIEW

File: dashboard/src/lib/porter-apps/values.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { z } from "zod";
+import { ServiceField, serviceNumberValidator, serviceStringValidator } from "./services";
+
+// Autoscaling
+export const autoscalingValidator = z.object({
+  minInstances: serviceNumberValidator,
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/values.ts:6]: also include enabled field?

File: dashboard/src/lib/porter-apps/values.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  maxInstances: serviceNumberValidator,
+  cpuThresholdPercent: serviceNumberValidator,
+  memoryThresholdPercent: serviceNumberValidator,
+});
+export type ClientAutoscaling = z.infer<typeof autoscalingValidator>;
+export type SerializedAutoscaling = {
+  minInstances: number;
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/values.ts:13]: enabled?

File: dashboard/src/lib/porter-apps/values.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  override,
+}: {
+  autoscaling?: SerializedAutoscaling;
+  override?: SerializedAutoscaling;
+}): ClientAutoscaling | undefined {
+  if (!autoscaling) {
+    return undefined;
+  }
+
+  return {
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/values.ts:45]: enabled?

File: dashboard/src/lib/porter-apps/values.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    memoryThresholdPercent: ServiceField.number(
+      autoscaling.memoryThresholdPercent,
+      override?.memoryThresholdPercent
+    ),
+  };
+}
+
+// Health Check
+export const healthcheckValidator = z.object({
+  httpPath: serviceStringValidator,
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/values.ts:67]: same comments about `enabled`

File: dashboard/src/main/home/app-dashboard/types/buildpack.ts [hunk 5]
```diff
@@ -1,32 +1,36 @@
...
-export type DetectedBuildpack = {
-    name: string;
-    builders: string[];
-    detected: Buildpack[];
-    others: Buildpack[];
-    buildConfig: BuildConfig;
-};
+import { z } from "zod";
+
+export const buildConfigSchema = z.object({
```
Comment (d-g-town) [inline dashboard/src/main/home/app-dashboard/types/buildpack.ts:3]: I'm not sure what the use case of this is: only difference between this and buildpackScehma is the array of buildpacks, but then we have an array of buildpackSchema in the `detectedBuildpackSchema`

Comment (ianedwards) [reply dashboard/src/main/home/app-dashboard/types/buildpack.ts:3]: I'll double check on this, but my concern is backwards compatibility with other components which all use these types, so I didn't get a chance to audit that as part of this PR

File: dashboard/src/lib/porter-apps/services.ts [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+      { type: "web" },
+      (config) =>
+        new Service({
+          ...service,
+          type: getServiceTypeEnumProto(config.type),
+          config: {
+            value: {
+              ...config,
+            },
+            case: "webConfig",
```
Comment (d-g-town) [inline dashboard/src/lib/porter-apps/services.ts:286]: `kind` or `type` feels better than case

Comment (ianedwards) [reply dashboard/src/lib/porter-apps/services.ts:286]: This is an autogenerated type from api-contracts

**Review by d-g-town — APPROVED** [review]

**Comment by d-g-town** [comment]
High-level non-blocking suggestions:
Remove all the `get`s from the beginning of methods
Add return types to all the methods in the method declaration

**Pull request merged**